---
title: "DE analysis- virus load as explanatory variable"
author: "Nurit"
date: "4/23/2020"
output: html_document
---


```{r libraries}
library(tidyverse)
library(DESeq2)
library(tximport)
library(rhdf5)
library(readr)
library(dplyr)
library(apeglm)
library(readr)
library(EnhancedVolcano)
library(ggplot2)
library(DEGreport)
library(autoplotly)
library(iSEE)
```

following read mapping using Kallisto, we wish to plot the bootstraps against the viral load as a continous variable. 
However, the sleuth package doesnt apply for continous variables,
so we need to import the data from kallisto and analyse it using DESeq 2:
 
## Analyzing RNA-seq data with DESeq2 (Love et al. 2014)
#### Standard workflow
[DESeq 2 tutorial](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html) 

***
If you have performed transcript quantification (with Salmon, *kallisto*, RSEM, etc.) you could import the data with tximport, which produces a list, and then you can use DESeqDataSetFromTximport().

> Import transcript-level estimates from kallisto using "tximport" (Soneson et al. 2015) [Import tutorial](https://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html#import_transcript-level_estimates) 

### first, make the auxillary table that describes the experimental design and the relationship between the kallisto directories and the samples, that is "exp_var" data frame
```{r echo=TRUE}
setwd("/Users/nuriteliash/OneDrive - OIST/Repos/varroa-virus-networks") 

# read the table with the libraries description:
ModTraitFac_66 <- read_csv("/Users/nuriteliash/OneDrive - OIST/Repos/varroa-virus-networks/module_trait_factors_66.csv")

# load the viruses load by library, for all 23 viruses
lnames <- load(file = "viruses_load.RData") 


# First, we locate the directory containing the files
dir <- file.path("/Users/nuriteliash/OneDrive - OIST/Repos/varroa-virus-networks/sleuth/kallistoAll")
list.files(dir)

# make a data frame of the viral loads of DWVa and VDV2 to each library (the explanatory variable). 
DWVa_VDV2_load <- viruses_load %>%
    filter(description %in% c("DWVa", "VDV2")) %>%
    column_to_rownames("description") %>%
  dplyr::select(-c("SRR5109825", "SRR5109827", "SRR533974" , "SRR3927496", "SRR8867385")) %>%
    t() %>%
    as.data.frame() %>%
    rownames_to_column("Library") 
  
# make the final "exp_var" table conatining the viral load, and the other covariates
exp_var <- left_join(DWVa_VDV2_load, ModTraitFac_66, by = "Library")

# now create a table named "sample", with the condition information (in our case DWVa or VDV2 viral load)  already be present as a column of the sample table samples:
samples <- exp_var %>%
  dplyr::select(run = Library, condition = DWVa, stage = mite_stage, host = bee_sp, lib_prep = Lib_treat)
# for VDV2 , just change the "condition = VDV2", then run the next steps the same

# Next we specify the path to the files using the appropriate columns of samples
files <- file.path(dir, samples$run, "abundance.h5")
names(files) <- samples$run

```


## DE for all genes in kallisto
```{r}
# We import the necessary quantification data for DESeq2 using the tximport function.and we read in a table that links transcripts to genes for this dataset.
tx2gene <- read_tsv("/Users/nuriteliash/OneDrive - OIST/Repos/varroa-virus-networks/data/gene2isoform.txt.gz", col_names = c("gene_id", "target_id"))
tx2gene <- tx2gene[,c(2,1)]
#then import the kallisto files:
txi <- tximport(files, type="kallisto", tx2gene=tx2gene)

# save the "txi" list
saveRDS(txi, file = "txi_allGenes_66.rds")

```


```{r}
#Finally, we can construct a DESeqDataSet from the txi object and sample information in "samples". the "condition" is the virus tpm (a continuos variable)
ddsTxi <- DESeqDataSetFromTximport(txi,
                                   colData = samples,
                                   design = ~ host + stage + lib_prep + condition)

# now we run the DESeq2 analysis:
dds <- DESeq(ddsTxi)
res <- results(dds)
res

# save the DE results of all genes
saveRDS(res, file = "DEresults_allGenes")
```



more analysis and visualisation:
```{r}
#We can order our results table by the smallest p value:
resOrdered <- res[order(res$pvalue),]

#We can summarize some basic tallies using the summary function.
summary(res)

#How many adjusted p-values were less than 0.1?
sum(res$padj < 0.1, na.rm=TRUE)

resultsNames(dds)

# in total, what genes have both (1)adjusted p-value<0.05, and (2) log2 fold change >1 (upregulation), or  log2 fold change < 0.5 (downregulation) in reponse to viral load.
DEgenes <- c(res$padj < 0.1, na.rm=TRUE)
SigGenes <- subset(res, padj<0.1)


# among the significant genes, how many are upregulated in reponse to viral load?
SigUpGenes <- subset(SigGenes,log2FoldChange > 0)
SigUpGenes
# among the significant genes, how many are downregulated in reponse to viral load?
SigDwnGenes <- subset(SigGenes,log2FoldChange < 0)
SigDwnGenes

# צריך להחליט למה זה טוב
#volcano plot the genes according to Log2FoldChange, 
EnhancedVolcano(res,
    lab = rownames(res),
    x = 'log2FoldChange',
    y = 'padj',
    xlim = c(-0.00001, 0.00001))

# another option
volcano <- ggplot(results, aes(x=log2FoldChange, y=-log(padj))) +
  geom_point()
volcano


#MA-plot
# In DESeq2, the function plotMA shows the log2 fold changes attributable to a given variable over the mean of normalized counts for all the samples in the DESeqDataSet. Points will be colored red if the adjusted p value is less than 0.1. Points which fall out of the window are plotted as open triangles pointing either up or down.
plotMA(res, ylim=c(-0.00001,0.00001))



# We provide the dds object and the name or number of the coefficient we want to shrink, where the number refers to the order of the coefficient as it appears in resultsNames(dds).
resultsNames(dds)

# Plot counts
# It can also be useful to examine the counts of reads for a single gene across the groups. A simple function for making this plot is plotCounts, which normalizes counts by sequencing depth and adds a pseudocount of 1/2 to allow for log scale plotting. The counts are grouped by the variables in intgroup, where more than one variable can be specified. Here we specify the gene which had the smallest p value from the results table created above. You can select the gene to plot by rowname or by numeric index.

#for the min gene:
d <- plotCounts(dds, gene=which.min(res$padj), intgroup="condition", 
                returnData=TRUE)

#for specific gene, e.g. "111243972"
d <- plotCounts(dds, gene="111243972", intgroup="condition", 
                returnData=TRUE)

# and plot it:
ggplot(d, aes(x=condition, y=count)) + 
  geom_point(position=position_jitter(w=0.1,h=0)) + 
  scale_y_log10(breaks=c(25,100,400))

# The runApp function launches the app in our browser.
app <- iSEE(dds)
shiny::runApp(app)


## more information:
#Information about which variables and tests were used can be found by calling the function mcols on the results object.
mcols(res)$description

#make a new dataframe with the DE genes:
#baseMean <- as.matrix(SigDwnGenes$baseMean, SigDwnGenes$log2FoldChange, SigDwnGenes$padj)
#baseMean <- rename(baseMean, "gene_id" = "SigDwnGenes$baseMean")

```

## looking at the differential expression of specific genes-modules 
```{r}
#first, make a dataframe from the DE anlaysis results for all 10,247 genes
results<- res %>%
  as.data.frame() %>%
  rownames_to_column("X1")
  
# then, make a list of specific genes , you wish to look at (for each module)
geneInfo <- read_csv("/Users/nuriteliash/OneDrive - OIST/Repos/varroa-virus-networks/Varroa-modules/geneInfo.csv")
```

looking at specific genes-modules differential expression ("magenta" module)
```{r}
# a list of the genes in "magenta" module:
magentaGene <- geneInfo %>% 
    filter(moduleColor == "magenta") %>%
    select(X1, moduleColor)

magentaGene$X1 <- as.character(magentaGene$X1)

# return all the counts in "results", of the genes in "magentaGene". 
DE_magenta <- semi_join(results, magentaGene, by = "X1") 


#We can order our results table by the smallest p value:
magenta_rdered <- DE_magenta[order(DE_magenta$pvalue),]


```

looking at specific genes-modules differential expression ("cyan" module)
```{r}
# a list of the genes in "cyan" module:
cyanGene <- geneInfo %>% 
    filter(moduleColor == "cyan") %>%
    select(X1, moduleColor)
cyanGene$X1 <- as.character(cyanGene$X1)

# return all the counts in "results", of the genes in "magentaGene". 
DE_cyan <- semi_join(results, cyanGene, by = "X1") 
  
#We can order our results table by the smallest p value:
cyan_rdered <- DE_cyan[order(DE_cyan$pvalue),]
```

looking at specific genes-modules differential expression ("turquoise" module)
```{r}
# a list of the genes in "turquoise" module:
turquoiseGene <- geneInfo %>% 
    filter(moduleColor == "turquoise") %>%
    select(X1, moduleColor)
turquoiseGene$X1 <- as.character(turquoiseGene$X1)
  
# return all the counts in "results", of the genes in "magentaGene". 
DE_turquoise <- semi_join(results, turquoiseGene, by = "X1") 
   #We can order our results table by the smallest p value:
turquoise_rdered <- DE_turquoise[order(DE_turquoise$pvalue),] 
```


###Exploring and exporting results

MA-plot
In DESeq2, the function plotMA shows the log2 fold changes attributable to a given variable over the mean of normalized counts for all the samples in the DESeqDataSet. Points will be colored red if the adjusted p value is less than 0.1. Points which fall out of the window are plotted as open triangles pointing either up or down.

```{r}
plotMA(res, ylim=c(-2,2))

```

Extracting transformed values
These transformation functions return an object of class DESeqTransform which is a subclass of RangedSummarizedExperiment. For ~20 samples, running on a newly created DESeqDataSet, rlog may take 30 seconds, while vst takes less than 1 second. The running times are shorter when using blind=FALSE and if the function DESeq has already been run, because then it is not necessary to re-estimate the dispersion values. The assay function is used to extract the matrix of normalized values.
```{r}
vsd <- vst(dds, blind=FALSE)
#rld <- rlog(dds, blind=FALSE)
head(assay(vsd), 3)

```

Principal component plot of the samples
Related to the distance matrix is the PCA plot, which shows the samples in the 2D plane spanned by their first two principal components. This type of plot is useful for visualizing the overall effect of experimental covariates and batch effects.
```{r}

plotPCA(vsd, intgroup=c("condition"))

``` 


--------------------------

## OLD - things i didnt use

#or, if we want to keep the transcripts, and not conect them the the gene (txOut = TRUE):
#txi.kallisto <- tximport(files, type = "kallisto", txOut = TRUE)
#head(txi.kallisto$counts)

#if using transcripts only, before constructing the DESeqDataSet, i will scale and center the virus tpm variable 
#using "scale" function on the "condition" (the virus tpm) in the "sample" dataframe.
#samples$condition <- scale(samples$condition, center = TRUE, scale = TRUE)

### the old command for DE using Sleuth:
```{r upload kallisto data, eval=FALSE, include=FALSE}
setwd("/Users/nuriteliash/OneDrive - OIST/Repos/varroa-virus-networks")

# The first step in a sleuth analysis is to specify where the kallisto results are stored. A variable is created for this purpose with
sample_id <- list.dirs(path = "sleuth/kallistoAll", full.names = FALSE) 
#sample_id <- sub("(.*).tsv.gz","\\1", sample_id)


#original code. אני לא בטוחה מה ההבדל בין הפקודות. אבל הפקודה המקורית מחזירה וקטור ריק.  
#sample_id <- dir(file.path("..", "results"))


# A list of paths to the kallisto results indexed by the sample IDs is collated with
kal_dirs <- file.path("sleuth/kallistoAll", sample_id)
#remove the first element (i dont know why, but the first element is an emtpy folder)
kal_dirs <- kal_dirs[-1]

#original code:
#kal_dirs <- file.path("..", "results", sample_id, "kallisto")

# The next step is to load an auxillary table that describes the experimental design and the relationship between the kallisto directories and the samples:
# read the table with the libraries description:
ModTraitFac_71 <- read.csv("/Users/nuriteliash/OneDrive - OIST/Repos/varroa-virus-networks/module_trait_factors_71.csv", stringsAsFactors=TRUE) 

# load the viruses load by library, for all 23 viruses
lnames <- load(file = "viruses_load.RData"); 

# make a data frame of the viral loads of DWVa and VDV2 to each library (the explanatory variable)
exp_var <- viruses_load %>%
    filter(description %in% c("DWVa", "VDV2")) %>%
    column_to_rownames("description") %>%
    t() %>%
    as.data.frame() %>%
    rownames_to_column("Library")
  

# join the two tables (rename the library as "sample" to match the next steps)
# **for DWVa**:
s2c <- left_join(exp_var, ModTraitFac_71 ,by = "Library") %>%
    dplyr::select(sample = Library, condition = DWVa)
# *there is a problem joining 4 libs, their factors doesnt join. i dont know why*


# Now the directories must be appended in a new column to the table describing the experiment. This column must be labeled path, otherwise sleuth will report an error. This is to ensure that samples can be associated with kallisto quantifications.
s2c <- dplyr::mutate(s2c, path = kal_dirs)

#It is important to check that the pairings are correct:
print(s2c)
```


