---
title: "GO term Varroa"
author: "Nurit"
date: "2/5/2021"
output: html_document
---


# GO Enrichment analysis for specific modules, and for the hub-genes

# Load libraries

```{r libraries}
library(dplyr)
library(tidyverse)
library(vegan)
library(DESeq2)
library(ggfortify)
library(WGCNA)
library(rmarkdown)
library("knitr") # for the markdown
library("kableExtra") # for creating a scrolling table
library("ggplot2") # for plotting 
library("ape") # for reading the phylogenetic tree and for mantel.test
library("Biostrings")
library("ggtree") # for plotting the tree
library("ggrepel") # for spreading text labels on the plot
library("scales") # for axis labels notation
library("GO.db")
library("reshape2")
library("RSQLite")
library("AnnotationDbi")
library("GSEABase")
library("GOstats")
library("maps") # for the map background
library("leaflet") #for the interactive maps
library("htmltools")
library("rgdal")
library("grid")
library("gridExtra")
library("GeneOverlap")
library(cluster)
```

- data provided in here [link](https://github.com/MaevaTecher/varroa-denovo-genomes/blob/master/data/Positive%20selection/Vdesselected1511.csv)
- codes provided in here [link](https://maevatecher.github.io/varroa-denovo-genomes/#go_term_enrichment_of_positively_selected_genes)
```{r eval=FALSE echo=TRUE}
#Preparing the GO frame
annot.vd <- read.csv("/Users/nuriteliash/Documents/GitHub/varroa-virus-networks/data/VdesGOready2.csv") 

annot.vd2 <- annot.vd %>%
  mutate(evidence = "IEA") %>%
  dplyr::select(go_id = GO.ids, evidence, gene = Gene.id)

head(annot.vd2)

goFrame.vd <-GOFrame(annot.vd2, organism = "Vd")
goAllFrame.vd <-GOAllFrame(goFrame.vd)
gsc.vd <-GeneSetCollection(goAllFrame.vd, setType = GOCollection())

universe.vd <- unique(annot.vd2$gene)
head(universe.vd)

# Load the expression and trait data saved in the first part 
load("/Users/nuriteliash/OneDrive - OIST/Repos/varroa-virus-networks-Local/results/varroa_virus-01-dataInput.RData")

# load the network info
load("/Users/nuriteliash/OneDrive - OIST/Repos/varroa-virus-networks-Local/results/Varroa_modules_networkConstruction-auto.RData")

# change "cyan" to the name of the desired module, in the first line: [moduleColors=="cyan"], and in the final "write.csv(file = "GO_term_enrichment_**cyan**BP.csv")
ME <- names(for_modules)[moduleColors=="magenta"]
ME_df <- data.frame(geneno = ME,Gene.id = ME)
ME_df$Gene.id <- paste("LOC", ME_df$Gene.id, sep="")

#or change "ME_df" to any table of sets of genes, that i wish to do GO-enrichment analysis:
# for exanple, Goterms for the hub-genes only, (hub_annot_53): genes.vd <- read.csv("hub_annot_53.csv")

genes.vd <- as.integer(unique(ME_df$geneno))
head(genes.vd)

params.vd <- GSEAGOHyperGParams(name = "Vd_GO_enrichment",
                                geneSetCollection = gsc.vd,
                                geneIds = genes.vd,
                                universeGeneIds = universe.vd,
                                ontology = "BP", # change with MF, CC to test all
                                pvalueCutoff = 0.05,
                                conditional = F,
                                testDirection = "over")

over.vd <- hyperGTest(params.vd)

#summary(over.vd)
GO_enrich.vd <- as.data.frame(summary(over.vd))
GO_enrich.vd %>% 
  arrange(Pvalue) %>% 
  write.csv(file = "/Users/nuriteliash/OneDrive - OIST/Repos/varroa-virus-networks-Local/results/GO_term_magenta.csv")

#load the go term enrichment analysis done using package ‘Category’
genesGO_0 <- read.csv("/Users/nuriteliash/OneDrive - OIST/Repos/varroa-virus-networks-Local/results/GO_term_magenta.csv")

# keep only highly significant terms
genesGO <- genesGO_0 %>%
  filter(Pvalue < 0.0000000001) 

#you can play with the number of terms by adjusting the "filter(Pvalue < 0.0000000001)"

#now plot it
g <- genesGO %>%
  mutate(logPv = log(Pvalue))

#or plot it based on the count of each term and the pvalue:
ggplot(g, aes(x = 0, y = Term, size = Count, color = logPv)) +
  geom_point(alpha = 0.9) +
  theme_classic() +
  ggtitle("Magenta GO-terms")
```