---
title: "overlap_modules"
author: "Nurit Eliash"
date: "2/4/2021"
output:
  html_document:
    theme: paper
    toc: yes
    toc_depth: 5
    toc_float:
      collapsed: yes
      df_print: paged
---
*NOTE* first run: 
"varroa virus networks.Rmd"
"varroa_modules".rmd  
"varroa_modules_females.rmd"

```{r data storage, eval=FALSE, include=FALSE}
#the working directory is in GitHb:
setwd("/Users/nuriteliash/Documents/GitHub/varroa-virus-networks")

#I save all the big outputs results in my local storage, on OneDrive:
"/Users/nuriteliash/OneDrive - OIST/Repos/varroa-virus-networks-Local/results"

knitr::opts_chunk$set(echo = TRUE)
```

```{r libraries, include=FALSE}
# load library
library("data.table")
library("dplyr")
library("tidyverse")
library("vegan")
library("DESeq2")
library("ggfortify")
library("WGCNA")
library("rmarkdown")
library("knitr") # for the markdown
library("kableExtra") # for creating a scrolling table
library("ggplot2") # for plotting 
library("ape") # for reading the phylogenetic tree and for mantel.test
library("Biostrings")
library("ggtree") # for plotting the tree
library("ggrepel") # for spreading text labels on the plot
library("scales") # for axis labels notation
library("GO.db")
library("reshape2")
library("RSQLite")
library("AnnotationDbi")
library("GSEABase")
library("GOstats")
library("maps") # for the map background
library("leaflet") #for the interactive maps
library("htmltools")
library("rgdal")
library("grid")
library("gridExtra")
library("GeneOverlap")
library("cluster")
library("ggpmisc")
library("DEGreport")
```

> Aim: detecting genes that appear in matching modules of the two analyses (using 66 and 57 libraries).

before we continue to explore the genes in Magenta module (in the 66 analysis), we need to be sure that they overlap those in the matching module in the female-only analysis (57 libraries).

```{r load data, include=FALSE}
# load the module network, gene expression and viral load data of the 66 libraries analysis:
names <- load("/Users/nuriteliash/OneDrive - OIST/Repos/varroa-virus-networks-Local/results/Varroa_modules_networkConstruction-auto.RData")
names
#load geneTraitSignificance (geneTraitSignificance), module membership (geneModuleMembership) and intramodular connectivity (Alldegrees1) for each gene:
names <- load(file = "/Users/nuriteliash/OneDrive - OIST/Repos/varroa-virus-networks-Local/results/IntraModCon66.RData")
names
names <- load(file = "/Users/nuriteliash/OneDrive - OIST/Repos/varroa-virus-networks-Local/results/geneTraitANDgeneMM_66.RData")
names
#load the genes tpm per library, and viral load:
names <- load(file = "/Users/nuriteliash/OneDrive - OIST/Repos/varroa-virus-networks-Local/results/varroa_virus-01-dataInput.RData")
names
#load the correlation data for the module-trait relationship:
names <- load("/Users/nuriteliash/OneDrive - OIST/Repos/varroa-virus-networks-Local/results/moduleTraitCor_66.RData")
names

# and of the 57 libraries analysis:
names <- load("/Users/nuriteliash/OneDrive - OIST/Repos/varroa-virus-networks-Local/results/Varroa_modules_networkConstruction-auto_57.RData")
names
names <- load(file = "/Users/nuriteliash/OneDrive - OIST/Repos/varroa-virus-networks-Local/results/IntraModCon57.RData")
names
names <- load(file = "/Users/nuriteliash/OneDrive - OIST/Repos/varroa-virus-networks-Local/results/geneTraitANDgeneMM_57.RData")
names
#load the genes tpm per library, and viral load:
names <- load(file = "/Users/nuriteliash/OneDrive - OIST/Repos/varroa-virus-networks-Local/results/varroa_virus-01-dataInput_57.RData")
names
#load the correlation data for the module-trait relationship:
names <- load("/Users/nuriteliash/OneDrive - OIST/Repos/varroa-virus-networks-Local/results/moduleTraitCor_57.RData")
names
```

# Finding matching modules (66 all libraries vs. 57 female libraries)
```{r, overlap modules, message=FALSE, warning=FALSE}
# overlap the two modules tables, to check if the modules contain the same genes
# to avoid confusion, i used the module colors for the 66 libs ("moduleColors" 15 modules + grey=genes with no module), and the modules numbers for the 57 libs ("moduleLabels_57" 14 modules + "zero" = genes with no module):
compareModules <- overlapTable(
    moduleColors, moduleLabels_57, 
    na.rm = FALSE, ignore = NULL, 
    levels1 = NULL, levels2 = NULL)
compareModules
```
using this function i was able to match modules from the analysis including all 66 libraries, and the analysis of females only (57 libraries). 
matching modules of the two analyses

66 all libraries                  | 57 female libraries:
----------------------------------|---------------------------
(module number 0)  MM.grey        | MM. (module number 0)
(module number 1)  MM.turquoise   | MM.turquoise (module number 1)
(module number 2)  MM.blue        | MM.red (module number 6)
(module number 3)  MM.brown       | MM.grey (module number 0)
(module number 4)  MM.yellow      | MM.black (module number 7)
(module number 5)  MM.green       | MM.yellow (module number 4)
(module number 6)  MM.red         | MM.brown (module number 3)
(module number 7)  MM.black       | MM.grey (module number 0)
(module number 8)  MM.pink        | MM.pink (module number 8)
(module number 9)  MM.magenta     | MM.brown (module number 3)
(module number 10) MM.purple      | MM.magenta (module number 9)
(module number 11) MM.greenyellow | MM.pink (module number 8)
(module number 12) MM.tan         | MM.pink (module number 8)
(module number 13) MM.salmon      | MM.yellow (module number 4)
(module number 14) MM.cyan        | MM.pink (module number 8)
(module number 15) MM.midnightblue| MM.tan (module number 12)
---

> Intersting modules:

according to "Module-Trait interaction", Magenta module looks interesting, as it have a negative correlation with DWVa, and positive correlation with VDV2.
the matching module of *Magenta (in the 66libs analysis)* is *black (in the 57libs analysis)*.
So we gonna overlap the black module from 57, to magenta genes, from 66 to get a table containing the overlapping genes, their geneTraitSignificance (for DWVa, ARV_2 and VDV2), geneModuleMembership, and Intramodular connectivity
```{r, make a table of the overlapping genes in Magenta module, message=FALSE, warning=FALSE}
geneInfoMagenta_66 = data.frame(
  module = moduleColors,
  geneTraitSignificance_66,
  geneModuleMembership_66,
  IntraModCon = Alldegrees1_66$kWithin) %>%
  dplyr::select(c(module, IntraModCon, MMmagenta, GS.DWVa, GS.VDV2, GS.ARV_2)) %>%
  dplyr::filter(module == "magenta") %>%
  rownames_to_column("genes")
  
geneInfoBlack_57 = data.frame(
  module = moduleColors_57,
  geneTraitSignificance_57,
  geneModuleMembership_57,
  IntraModCon = Alldegrees1_57$kWithin) %>%
  dplyr::select(c(module, IntraModCon, MMblack, GS.DWVa, GS.VDV2, GS.ARV_2)) %>%
  dplyr::filter(module%in% c("black")) %>%
  rownames_to_column("genes")

# make a table of the overlapping genes 
overlap <- inner_join(geneInfoMagenta_66, geneInfoBlack_57, by = "genes")
 #change the columns names to specify from which analysis it was taken (66libs or 57libs)
 overlap <- overlap %>%
   dplyr::rename(module_66 = module.x,
                 module_57 = module.y,
                 IntraModCon_66 = IntraModCon.x,
                 IntraModCon_57 = IntraModCon.y,
                 GS.DWVa_66 = GS.DWVa.x,
                 GS.DWVa_57 = GS.DWVa.y,
                 GS.VDV2_66 = GS.VDV2.x,
                 GS.VDV2_57 = GS.VDV2.y,
                 GS.ARV_2_66 = GS.ARV_2.x,
                 GS.ARV_2_57 = GS.ARV_2.y)
nrow(overlap)  
overlap
# we have 40 overlapping genes in the magenat and balck modules, as the "overlapTable()" function computed.

# add gene annotation:
# load the annotation file:
annot_varroa <- read_csv("/Users/nuriteliash/Documents/GitHub/varroa-virus-networks/data/annot_varroa.csv", col_names = TRUE, )

# remove the "LOC" from the gene name
annot_varroa$Locus <- str_replace(annot_varroa$Locus, "LOC", '')
# and change the col name to "genes", so it will the same as in the "overlap" table
colnames(annot_varroa)[which(names(annot_varroa) == "Locus")] <- "genes"
head(annot_varroa)

#now join by "genes" name (INNER JOIN: returns rows when there is a match in both tables)
overlap_annot <- inner_join(overlap, annot_varroa, by = "genes") %>% 
  dplyr::select(c(genes, module_66, module_57,
                 IntraModCon_66, 
                 IntraModCon_57, 
                 GS.DWVa_66,  
                 GS.DWVa_57, 
                 GS.VDV2_66,
                 GS.VDV2_57, 
                 GS.ARV_2_66,
                 GS.ARV_2_57, Name, Accession, 'Protein product', Length, 'Protein Name'))

# and filter out duplicated genes
overlap_annot <-overlap_annot[!duplicated(overlap_annot[,'genes']),] 

# how many genes overlap between Magenta module of the 66 libs, and the Black module of the 57 modules?
nrow(overlap_annot)

write_csv(x = as.data.frame(overlap_annot), "results/overlap_annot.csv")
```

we have 40 overlapping genes between the two analyses, out of 86 Magenta module genes (66 libraries), and out of 328 genes in Black module (57 libraries). 

***
We also want to check the genes that were silenced, which were mistakenly chosen from the "old-hubgenes"

# check the old-hubgenes - and specifically the silenced genes.
1. do Magenta module overlap with the old-hubgenes?
2. to which module the silenced genes belong to?

## (1) do Magenta module overlap with the old-hubgenes?
```{r, overlapping Magenta and old-hubgenes, message=FALSE, warning=FALSE}
#load the former hub-genes:
hub_old <- read_csv("/Users/nuriteliash/Documents/GitHub/varroa-virus-networks/data/hub_annot_old.csv", col_names = TRUE, ) 
hub_old$genes <- as.character(hub_old$genes)

# how many of the hub_old genes exist in the magenta module?
#return all rows from x, and all columns from x and y. 
f <- left_join(geneInfoMagenta_66, hub_old, by = "genes")

sum(!is.na(f$virus.x))

# only 3 genes of magenta module appear in the former hubgenes analysis
# so to what module these 53 old-hubgenes belong to?

#make a data.frame of all genes with modules:
all.genes.66 <- data.frame(
  module = moduleColors,
  geneTraitSignificance_66,
  GSPadjust_66,
  geneModuleMembership_66,
  MMPadjust_66,
  IntraModCon = Alldegrees1_66$kWithin) %>%
  dplyr::select(c(module, IntraModCon, GS.DWVa, pdj.GS.DWVa, GS.VDV2, pdj.GS.VDV2, GS.ARV_2, pdj.GS.ARV_2, MMpurple, padj.MMpurple, MMpink, padj.MMpink, MMgreen, padj.MMgreen)) %>%
  rownames_to_column("genes")

left_join(hub_old, all.genes.66, by = "genes")

all.genes.57 <- data.frame(
  module = moduleColors_57,
  geneTraitSignificance_57,
  GSPadjust_57,
  geneModuleMembership_57,
  MMPadjust_57,
  IntraModCon = Alldegrees1_57$kWithin) %>%
  dplyr::select(c(module, IntraModCon, GS.DWVa, pdj.GS.DWVa, GS.VDV2, pdj.GS.VDV2, GS.ARV_2, pdj.GS.ARV_2, MMsalmon, padj.MMsalmon)) %>%
  rownames_to_column("genes")

left_join(hub_old, all.genes.57, by = "genes")
```
> Conclusions:

only 3 genes of magenta module appear in the former hubgenes analysis
in the 66 libs, most old-hubgenes belong to the purple module 
in the 57 libs, most old-hubgenes belong to the salmon module

> "1. do Magenta module overlap with the old-hubgenes?" 

> No: I made a mistake, and only 3 of the old-hubgenes belong to the actually interesting module, Magenta  

what about the silenced genes? 

## (2) to which module the silenced genes belong to?
```{r, modules of silenced-genes, message=FALSE, warning=FALSE}
# prepare a table describing each gene number, short-name and full annotation, of the silenced genes:
genes <- as.factor(c("111250594", "111244103","111244832", "111248360", "111245345", "111248674", "111251059", "111245371", "111244631", "111252566", "111249664"))
annot <- as.factor(c("Sarcalumenin", "glycerol-3-phosphate dehydrogenase", "Calmodulin", "Cuticle-protein8","Cuticle-protein-14", "Glutamate-receptor-3", "Glutamate-gated-chloride-channel-subunit-beta", "Organic-cation-transporter-protein", "Twitchin", "Annulin", "Defense-protein-Hdd11"))
ShortName <- as.factor(c("Sar", "Gly","clmd", "CuP8", "CuP14", "Glut", "Chl", "Trans", "Twitch", "Anl", "HDD"))
silenced.genes <- data.frame(genes, ShortName, annot)

# to what module the silenced genes belong to?
silenced_overlap.66 <- left_join(silenced.genes, all.genes.66, by = "genes")
silenced_overlap.57 <- left_join(silenced.genes, all.genes.57, by = "genes")
silenced_overlap.66
silenced_overlap.57

write_csv(x = as.data.frame(silenced_overlap.66), "results/silenced_genes_info.csv")
```
* None of the 11 silenced-genes belong to the Magenta module (8 belong to purple module, 2 to pink and 1 to green). 
* all have negative interaction with DWVa. 
 
 ***

but we have no choice now, as we already silenced these 11 genes. 
so we need to know if the silencing effect on the viral load suit the prediction of the module-virus interaction.
for that, we need to know: 

*(1) the prediction of the network analysis for the virus-module interaction* (virus tpm - module eigengene), which is actually “module-trait interaction”
*(2) the connectivity of each gene to their module* (eigengene - gene tpm)
*(3) the specific virus-gene interaction* (virus tpm - gene tpm), which is GeneSignificance 

# Pairwise correlations of Viral load, Module eigengenes and Gene expression:

## (1) the prediction of the network analysis for the virus-module interaction
virus - module (virus tpm - module eigengene). 

### (1a) Quantifying module–trait associations, for the 66 libraries
```{r, from the chunk of WGCNA 3.a Quantifying module–trait associations: 66, message=FALSE, warning=FALSE}
# first Recalculate MEs with color labels 
MEs0 = moduleEigengenes(for_modules, moduleColors)$eigengenes
MEs = orderMEs(MEs0)

sizeGrWindow(10,6)
# Will display correlations and their p-values
textMatrix = paste(signif(moduleTraitCor_66, 2), "\n(",
                   signif(moduleTraitPadjust_66, 1), ")", sep = "");
dim(textMatrix) = dim(moduleTraitCor_66)
par(mar = c(6, 8.5, 3, 3));
# Display the correlation values within a heatmap plot
labeledHeatmap(Matrix = moduleTraitCor_66,
               xLabels = names(b_viruses_load),
               yLabels = names(MEs),
               ySymbols = names(MEs),
               colorLabels = FALSE,
               colors = blueWhiteRed(50),
               textMatrix = textMatrix,
               setStdMargins = FALSE,
               cex.text = 0.7,
               zlim = c(-1,1),
               main = paste("Varroa Module-viruses relationships, all 66 libraries"))
```

### (1b) Quantifying module–trait associations, for the 57 libraries
```{r, from the chunk of WGCNA 3.a Quantifying module–trait associations: 57, message=FALSE, warning=FALSE}

## (1) virus > module (virus tpm > module eigengene). for the 66 libraries
# first Recalculate MEs with color labels 
MEs0 = moduleEigengenes(for_modules_female, moduleColors_57)$eigengenes
MEs_57 = orderMEs(MEs0)

sizeGrWindow(10,6)
# Will display correlations and their p-values
textMatrix = paste(signif(moduleTraitCor_57, 2), "\n(",
                   signif(moduleTraitPadjust_57, 1), ")", sep = "");
dim(textMatrix) = dim(moduleTraitCor_57)
par(mar = c(6, 8.5, 3, 3));
# Display the correlation values within a heatmap plot
labeledHeatmap(Matrix = moduleTraitCor_57,
               xLabels = names(b_viruses_load_57),
               yLabels = names(MEs_57),
               ySymbols = names(MEs_57),
               colorLabels = FALSE,
               colors = blueWhiteRed(50),
               textMatrix = textMatrix,
               setStdMargins = FALSE,
               cex.text = 0.7,
               zlim = c(-1,1),
               main = paste("Varroa Module-viruses relationships females only libraries"))
```

## (2) the connectivity of each gene to their module
Module - gene correlation (eigengene - gene tpm)
```{r,  Module - gene correlation, emessage=FALSE, warning=FALSE}
# make expression (tpm) of all silenced genes
genes <- as.factor(c("111250594", "111244103","111244832", "111248360", "111245345", "111248674", "111251059", "111245371", "111244631", "111252566", "111249664"))

silenced_exp <- for_modules %>%
  dplyr::select(genes)  %>%
  rownames_to_column("Library")

# make table of eigengens of the silenced genes
# first Recalculate MEs with color labels 
MEs0 = moduleEigengenes(for_modules, moduleColors)$eigengenes
MEs = orderMEs(MEs0)

Sil_MEs <- as.character(c("MEpurple", "MEgreen", "MEpink"))

silenced_MEs <- MEs %>%
  dplyr::select(all_of(Sil_MEs)) %>%
  rownames_to_column("Library")
  
# correlate gene expression (for each of the silenced genes) with eigengenes (of the gene to its module). 
geneMEcor <- as.matrix(cor(column_to_rownames(silenced_exp, "Library"), column_to_rownames(silenced_MEs, "Library"), "p"))

plot <- left_join(silenced_MEs, silenced_exp, by = "Library")

# plot the correlation of each gene and its MEs
GlyPurple <- ggplot(plot, aes(x=plot$"111244103", y=MEpurple)) + 
         geom_point() +
  geom_cor(method = "pearson", xpos = 1) +
  xlab("Gly gene expression (tpm)") + 
  ylab("Purple MEs)") +
    ggtitle("Gly expression correlation\n to Purple ME") +
      theme(
        plot.title = element_text(color="black", size=14, face="bold"),
        axis.title.x = element_text(color="black", size=14),
        axis.title.y = element_text(color="black", size=14))

SarPink <- ggplot(plot, aes(x=plot$"111250594", y=MEpink)) + 
         geom_point() +
  geom_cor(method = "pearson", xpos = 1) +
  xlab("Sar gene expression (tpm)") + 
  ylab("Pink MEs)") +
    ggtitle("Sar expression correlation\n to Pink ME") +
      theme(
        plot.title = element_text(color="black", size=14, face="bold"),
        axis.title.x = element_text(color="black", size=14),
        axis.title.y = element_text(color="black", size=14))

clmdPurple <- ggplot(plot, aes(x=plot$"111244832", y=MEpurple)) + 
         geom_point() +
  geom_cor(method = "pearson", xpos = 1) +
  xlab("clmd gene expression (tpm)") + 
  ylab("Purple MEs)") +
    ggtitle("clmd expression correlation\n to Purple ME") +
      theme(
        plot.title = element_text(color="black", size=14, face="bold"),
        axis.title.x = element_text(color="black", size=14),
        axis.title.y = element_text(color="black", size=14))

CuP8Purple <- ggplot(plot, aes(x=plot$"111248360", y=MEpurple)) + 
         geom_point() +
  geom_cor(method = "pearson", xpos = 1) +
  xlab("CuP8 gene expression (tpm)") + 
  ylab("Purple MEs)") +
    ggtitle("CuP8 expression correlation\n to Purple ME") +
      theme(
        plot.title = element_text(color="black", size=14, face="bold"),
        axis.title.x = element_text(color="black", size=14),
        axis.title.y = element_text(color="black", size=14))

CuP14Purple <- ggplot(plot, aes(x=plot$"111245345", y=MEpurple)) + 
         geom_point() +
  geom_cor(method = "pearson", xpos = 1) +
  xlab("CuP14 gene expression (tpm)") + 
  ylab("Purple MEs)") +
    ggtitle("CuP14 expression correlation\n to Purple ME") +
      theme(
        plot.title = element_text(color="black", size=14, face="bold"),
        axis.title.x = element_text(color="black", size=14),
        axis.title.y = element_text(color="black", size=14))

GlutPurple <- ggplot(plot, aes(x=plot$"111248674", y=MEpurple)) + 
         geom_point() +
  geom_cor(method = "pearson", xpos = 1) +
  xlab("Glut gene expression (tpm)") + 
  ylab("Purple MEs)") +
    ggtitle("Glut expression correlation\n to Purple ME") +
      theme(
        plot.title = element_text(color="black", size=14, face="bold"),
        axis.title.x = element_text(color="black", size=14),
        axis.title.y = element_text(color="black", size=14))

ChlPurple <- ggplot(plot, aes(x=plot$"111251059", y=MEpurple)) + 
         geom_point() +
  geom_cor(method = "pearson", xpos = 1) +
  xlab("Chl gene expression (tpm)") + 
  ylab("Purple MEs)") +
    ggtitle("Chl expression correlation\n to Purple ME") +
      theme(
        plot.title = element_text(color="black", size=14, face="bold"),
        axis.title.x = element_text(color="black", size=14),
        axis.title.y = element_text(color="black", size=14))


TwitchPurple <- ggplot(plot, aes(x=plot$"111244631", y=MEpurple)) + 
         geom_point() +
  geom_cor(method = "pearson", xpos = 1) +
  xlab("Twitch gene expression (tpm)") + 
  ylab("Purple MEs)") +
    ggtitle("Twitch expression correlation\n to Purple ME") +
      theme(
        plot.title = element_text(color="black", size=14, face="bold"),
        axis.title.x = element_text(color="black", size=14),
        axis.title.y = element_text(color="black", size=14))

HDDPurple <- ggplot(plot, aes(x=plot$"111249664", y=MEpurple)) + 
         geom_point() +
  geom_cor(method = "pearson", xpos = 1) +
  xlab("HDD gene expression (tpm)") + 
  ylab("Purple MEs)") +
    ggtitle("HDD expression correlation\n to Purple ME") +
      theme(
        plot.title = element_text(color="black", size=14, face="bold"),
        axis.title.x = element_text(color="black", size=14),
        axis.title.y = element_text(color="black", size=14))

AnlPink <- ggplot(plot, aes(x=plot$"111252566", y=MEpink)) + 
         geom_point() +
  geom_cor(method = "pearson", xpos = 1) +
  xlab("Anl gene expression (tpm)") + 
  ylab("pink MEs)") +
    ggtitle("Anl expression correlation\n to pink ME") +
      theme(
        plot.title = element_text(color="black", size=14, face="bold"),
        axis.title.x = element_text(color="black", size=14),
        axis.title.y = element_text(color="black", size=14))

TransGreen <- ggplot(plot, aes(x=plot$"111245371", y=MEgreen)) + 
         geom_point() +
  geom_cor(method = "pearson", xpos = 1) +
  xlab("Trans gene expression (tpm)") + 
  ylab("Green MEs)") +
    ggtitle("Trans expression correlation\n to Green ME") +
      theme(
        plot.title = element_text(color="black", size=14, face="bold"),
        axis.title.x = element_text(color="black", size=14),
        axis.title.y = element_text(color="black", size=14))


# all purple-module genes:
sizeGrWindow(20,10)
grid.arrange(GlyPurple, clmdPurple, CuP8Purple, CuP14Purple, GlutPurple, ChlPurple,  TwitchPurple, HDDPurple, nrow = 2)
# all pink-module genes:
grid.arrange(SarPink, AnlPink, nrow = 1)
# all green-module genes:
grid.arrange(TransGreen, nrow = 1)

# maybe better to make in a heatmap?
```

```{r eval=FALSE, include=FALSE}
#i tried to make the former chunk  in a loop:
------
genesNumb = as.character(silenced.genes$genes)
genesName = as.character(silenced.genes$ShortName)

sizeGrWindow(9,6)
par(mfrow=c(2,as.integer(0.5+length(genesNumb)/2)))
par(mar = c(4,5,3,1))
for (i in c(1:length(genesNumb)))
{
  whichgene=genesNumb[[i]];
  restrict1 = (genesNumb==whichgene);
  verboseScatterplot(plot$MEpink[restrict1],
                     geneTraitSignificance_66$GS.VDV2[restrict1], col=moduleColors[restrict1],
                     main=whichmodule,
                     xlab = "Connectivity", ylab = "Gene Significance", abline = TRUE)
}
i = 1
```

## (3) the specific virus-gene interaction. 
Virus > gene (virus tpm > gene tpm), which is GeneSignificance. 
```{r, virus-gene correlation, emessage=FALSE, warning=FALSE}
# upload the viruses load
viruses_load <- read_rds("/Users/nuriteliash/OneDrive - OIST/Repos/varroa-virus-networks-Local/results/viruses_load.RData")
viruses <- viruses_load %>%
  column_to_rownames("description") %>%
  transposeBigData() %>%
   dplyr::select(c("DWVa", "VDV2", "ARV_2")) %>%
   rownames_to_column("Library")
#%>% add_column(logDWVa = log(viruses$DWVa))

# make a correlation data frame between the 3 viruses and the 11 genes:
# put back the library as row_name
silenced_exp_1 <- silenced_exp %>%
  setnames(old = as.vector(genes), new = as.vector(ShortName)) %>%
  column_to_rownames("Library")

viruses_1 <- viruses %>%
  column_to_rownames("Library")

# Define numbers of genes and samples
nSamples = nrow(silenced_exp_1)

geneVirusCor <- cor(silenced_exp_1, viruses_1, "p")
geneVirusPvalue <- corPvalueStudent(geneVirusCor, nSamples) %>%
  as.data.frame() 

### Controlling the false discovery rate: Benjamini–Hochberg procedure ###
# using p.adjust function, for all comparisons, 11 genes and 3 viruses (m=33). 

#  "gather" all the p-values, so they will apear in one column
longer_Pvalue <- geneVirusPvalue %>% 
  rownames_to_column("genes") %>%
  gather("virus", "pvalue", -genes)

# now calculate the p.adjust for each p-value 
Padjust <- p.adjust(longer_Pvalue$pvalue, method = "fdr")

# and add the column of adjusted pvalues
Padjust <- add_column(longer_Pvalue, Padjust)

# now spread it back
geneVirusPadjust <- Padjust %>% 
  dplyr::select(-pvalue) %>% 
  group_by(virus) %>%
  pivot_wider(names_from = virus, values_from = Padjust)
geneVirusPadjust <- column_to_rownames(geneVirusPadjust, "genes")

# before correlating, make the two dataframes into "matrices"
geneVirusCor <- as.matrix(geneVirusCor)
geneVirusPadjust <- as.matrix(geneVirusPadjust)

# make a heat map of the correlation coefficients:

sizeGrWindow(10,6)
# Will display correlations and their p-values
textMatrix = paste(signif(geneVirusCor, 2), "\n(",
                   signif(geneVirusPadjust, 1), ")", sep = "");
dim(textMatrix) = dim(geneVirusCor)
par(mar = c(6, 8.5, 3, 3));
# Display the correlation values within a heatmap plot
labeledHeatmap(Matrix = geneVirusCor,
               xLabels = names(viruses_1),
               yLabels = names(silenced_exp_1),
               ySymbols = names(silenced_exp_1),
               colorLabels = FALSE,
               colors = blueWhiteRed(50),
               textMatrix = textMatrix,
               setStdMargins = FALSE,
               cex.text = 0.7,
               zlim = c(-1,1),
               main = paste("Varroa genes-viruses relationships"))
```

```{r, another option - plot the correlation of each gene-virus, eval=FALSE, include=FALSE}
# correlate gene expression (for each of the silenced genes) with viral load (of the three viruses)
plot1 <- left_join(viruses, silenced_exp, by = "Library") %>% 
  column_to_rownames("Library")

SarDWVa <- ggplot(plot1, aes(x=`111250594`, y=DWVa)) + 
         geom_point() +
  geom_cor(method = "pearson", xpos = 1) +
  xlab("Sar gene expression (tpm)") + ylab("DWVa viral load (log(tpm))") +
    ggtitle("Sar expression correlation to DWVa viral load") +
theme(
plot.title = element_text(color="black", size=14, face="bold"),
axis.title.x = element_text(color="black", size=14),
axis.title.y = element_text(color="black", size=14))
  
GlyDWVa <- ggplot(plot1, aes(x=`111244103`, y=DWVa)) + 
         geom_point() +
  geom_cor(method = "pearson", xpos = 1) +
  xlab("Gly gene expression (tpm)") + ylab("DWVa viral load (log(tpm))") +
    ggtitle("Gly expression correlation to DWVa viral load") +
theme(
plot.title = element_text(color="black", size=14, face="bold"),
axis.title.x = element_text(color="black", size=14),
axis.title.y = element_text(color="black", size=14))
```


### check correaltion between the overlapping genes connectivity of the two analyses:
```{r eval=FALSE, include=FALSE}
#### magenta (66) > purple (57) 
m <- overlap_annot %>% 
  dplyr::filter(module_57 == "purple")
my.formula <- y ~ x

# Calculating Pearson's product-moment correlation
 cor <- cor.test(m$IntraModCon_66, m$IntraModCon_57, method = "pearson", conf.level = 0.95)
 p.value <- cor$p.value
 r.squared <- cor$estimate
 
p <- ggplot(data = m, aes(x = IntraModCon_66, y = IntraModCon_57)) +
   geom_smooth(method = "lm", se=FALSE, color="black", formula = my.formula) +
   stat_poly_eq(formula = my.formula,
                aes(label = ..eq.label..), 
                parse = TRUE) + 
   labs(x = "magenta_66", y = "purple_57") + 
   ggtitle("magenta_purple Intramodular connectivity correlation") +
   geom_point() +
   geom_text(x=10, y=15, label=  paste("r = ", r.squared, sep="")) +
   geom_text(x=10, y=14, label=  paste("Pvalue = ", p.value, sep="")) + 
   theme_classic()
p

#### magenta (66) > black (57) 


m <- overlap_annot %>% 
  dplyr::filter(module_57 == "black")
my.formula <- y ~ x

# Calculating Pearson's product-moment correlation
 cor <- cor.test(m$IntraModCon_66, m$IntraModCon_57, method = "pearson", conf.level = 0.95)
 p.value <- cor$p.value
 r.squared <- cor$estimate
 
p <- ggplot(data = m, aes(x = IntraModCon_66, y = IntraModCon_57)) +
   geom_smooth(method = "lm", se=FALSE, color="black", formula = my.formula) +
   stat_poly_eq(formula = my.formula,
                aes(label = ..eq.label..), 
                parse = TRUE) + 
   labs(x = "magenta_66", y = "black_57") + 
   ggtitle("magenta_black Intramodular connectivity correlation") +
   geom_point() +
   geom_text(x=7, y=50, label=  paste("r = ", r.squared, sep="")) +
   geom_text(x=7, y=45, label=  paste("Pvalue = ", p.value, sep="")) + 
   theme_classic()
p
```

#### the intramodular connectivity of the genes in magenta module is correlated to that of its matching module (black), in the 57 libraries, but not to the purple module.

### check correaltion between the overlapping genes geneTraitSignificance of the two analyses, for DWVa virus
```{r eval=FALSE, include=FALSE}
#### magenta (66) > purple (57) 
m <- overlap_annot %>% 
  dplyr::filter(module_57 == "purple")
my.formula <- y ~ x

# Calculating Pearson's product-moment correlation
 cor <- cor.test(m$GS.DWVa_66, m$GS.DWVa_57, method = "pearson", conf.level = 0.95)
 p.value <- cor$p.value
 r.squared <- cor$estimate
 
p <- ggplot(data = m, aes(x = GS.DWVa_66, y = GS.DWVa_57)) +
   geom_smooth(method = "lm", se=FALSE, color="black", formula = my.formula) +
   stat_poly_eq(formula = my.formula,
                aes(label = ..eq.label..), 
                parse = TRUE) + 
   labs(x = "magenta_66", y = "purple_57") + 
   ggtitle("magenta_purple DWVa-Gene-significane correlation") +
   geom_point() +
   geom_text(x=-0.6, y=-0.3, label=  paste("r = ", r.squared, sep="")) +
   geom_text(x=-0.6, y=-0.35, label=  paste("Pvalue = ", p.value, sep="")) + 
   theme_classic()
p

#### magenta (66) > black (57) 


m <- overlap_annot %>% 
  dplyr::filter(module_57 == "black")
my.formula <- y ~ x

# Calculating Pearson's product-moment correlation
 cor <- cor.test(m$GS.DWVa_66, m$GS.DWVa_57, method = "pearson", conf.level = 0.95)
 p.value <- cor$p.value
 r.squared <- cor$estimate
 
p <- ggplot(data = m, aes(x = GS.DWVa_66, y = GS.DWVa_57)) +
   geom_smooth(method = "lm", se=FALSE, color="black", formula = my.formula) +
   stat_poly_eq(formula = my.formula,
                aes(label = ..eq.label..), 
                parse = TRUE) + 
   labs(x = "magenta_66", y = "purple_57") + 
   ggtitle("magenta_black DWVa-Gene-significane correlation") +
   geom_point() +
   geom_text(x=-0.6, y=-0.3, label=  paste("r = ", r.squared, sep="")) +
   geom_text(x=-0.6, y=-0.35, label=  paste("Pvalue = ", p.value, sep="")) + 
   theme_classic()
p
```

### check correaltion between the overlapping genes geneTraitSignificance of the two analyses, for VDV2 virus
```{r eval=FALSE, include=FALSE}
#### magenta (66) > purple (57) 
m <- overlap_annot %>% 
  dplyr::filter(module_57 == "purple")
my.formula <- y ~ x

# Calculating Pearson's product-moment correlation
 cor <- cor.test(m$GS.VDV2_66, m$GS.VDV2_57, method = "pearson", conf.level = 0.95)
 p.value <- cor$p.value
 r.squared <- cor$estimate
 
p <- ggplot(data = m, aes(x = GS.VDV2_66, y = GS.VDV2_57)) +
   geom_smooth(method = "lm", se=FALSE, color="black", formula = my.formula) +
   stat_poly_eq(formula = my.formula,
                aes(label = ..eq.label..), 
                parse = TRUE) + 
   labs(x = "magenta_66", y = "purple_57") + 
   ggtitle("magenta_purple VDV2-Gene-significane correlation") +
   geom_point() +
   geom_text(x=0.5, y=0.35, label=  paste("r = ", r.squared, sep="")) +
   geom_text(x=0.5, y=0.3, label=  paste("Pvalue = ", p.value, sep="")) + 
   theme_classic()
p

#### magenta (66) > black (57) 


m <- overlap_annot %>% 
  dplyr::filter(module_57 == "black")
my.formula <- y ~ x

# Calculating Pearson's product-moment correlation
 cor <- cor.test(m$GS.VDV2_66, m$GS.VDV2_57, method = "pearson", conf.level = 0.95)
 p.value <- cor$p.value
 r.squared <- cor$estimate
 
p <- ggplot(data = m, aes(x = GS.VDV2_66, y = GS.VDV2_57)) +
   geom_smooth(method = "lm", se=FALSE, color="black", formula = my.formula) +
   stat_poly_eq(formula = my.formula,
                aes(label = ..eq.label..), 
                parse = TRUE) + 
   labs(x = "magenta_66", y = "purple_57") + 
   ggtitle("magenta_black VDV2-Gene-significane correlation") +
   geom_point() +
   geom_text(x=0.5, y=0.8, label=  paste("r = ", r.squared, sep="")) +
   geom_text(x=0.5, y=0.75, label=  paste("Pvalue = ", p.value, sep="")) + 
   theme_classic()
p
```

### for both DWVa and VDV2 there is high correlation between the matching modules