---
title: "Varroa-virus meta-transcriptomic analysis"
author: "Sasha Mikheyev and Nurit Eliash"
date: "1/22/2020"
output:
  html_document:
    theme: paper
    toc: yes
    toc_depth: 5
    toc_float:
      collapsed: yes
      df_print: paged
---
this analysis is on the 66 libraries (i removed the 5 outlirered libraries from the "data/kallisto" folder in GitHub)
the working directory is in GitHb:
"/Users/nuriteliash/Documents/GitHub/varroa-virus-networks"

I save all the small outputs results in the "result" folder:
"/Users/nuriteliash/Documents/GitHub/varroa-virus-networks/results"

I save all the big outputs results in my local storage:
"/Users/nuriteliash/OneDrive - OIST/Repos/varroa-virus-networks-Local/results"

# Load libraries
```{r load library, message=FALSE, warning=FALSE, echo=FALSE}
library("dplyr")
library("tidyverse")
library("vegan")
library("DESeq2")
library("ggfortify")
library("WGCNA")
library("rmarkdown")
library("knitr") # for the markdown
library("kableExtra") # for creating a scrolling table
library("ggplot2") # for plotting 
library("ape") # for mantel.test
library("Biostrings")
library("ggrepel") # for spreading text labels on the plot
library("scales") # for axis labels notation
library("GO.db") # for GO term annotation 
library("reshape2")
library("RSQLite")
library("AnnotationDbi") # for GO term annotation 
library("GSEABase")
library("GOstats")
library("maps") # for the map background
library("htmltools")
library("rgdal")
library("grid")
library("gridExtra")
library("GeneOverlap") # for making the overlapping genes
library("cluster")
library("rmdformats")
library("corrplot")
library("viridis")
library("hrbrthemes")
library("ggthemes")
library("RColorBrewer")
library("naniar")
```

```{r}
knitr::opts_chunk$set(echo = TRUE)
setwd("/Users/nuriteliash/Documents/GitHub/varroa-virus-networks")
```

In this study we studied virus-virus and vector-virus interactions by two approaches: (1) meta-transcriptomic and (2) gene-silencing experiments. In the meta-transcriptomic analysis we looked at the overall viral landscape in the different libraries, detected vector modules (co-expressing genes using gene-network analysis), and correlated the viruses’ abundance to the vector modules. Last, we tested if the virus abundance matrix can predict the virus-vector interaction.

## Data selection
We searched for the term “varroa” in the SRA databases (NCBI, January 2020) with the following filters:
 *  _Source_: "RNA"
 *  _Library Strategy_: “RNA-seq” (Random sequencing of whole transcriptome)
 *  _Library Source_: “TRANSCRIPTOMIC”, “VIRAL RNA”
 *  Only whole mites (not partial organs)
 
## Viral abundance analysis and filtering data set




## Load data 
The input data frame "df" contains 35,669 isoforms of 66 varroa SRA libraries, downloaded from NCBI, using the 

after mapping the RNAseq reads to viruses and varroa genes using kallisto (____ref___), we gonna read and save the mapped reads and their tpm
```{r eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}
# First make a function for making the isoforms (mapped reads) into a data frame, where each column is a SAR library, and each row is an isoform. The cells contain the isoform TPM.  
read_kallisto <- function(filename) {
  sampleName <- sub("data/kallisto/(.*).tsv.gz","\\1", filename)
  return(read_tsv(filename) %>%
           dplyr::select(!!sampleName := tpm))
}

# now make the data frame 'df': 
df_71 <- list.files(path = "data/kallisto", full.names = TRUE) %>% 
  lapply(read_kallisto) %>% 
  bind_cols() 

# add a column "target_id" with the isoform/virus ID
df_71$target_id <- list.files(path = "data/kallisto", full.names = TRUE)[1] %>% 
  read_tsv() %>% 
  dplyr::select(target_id) %>% 
  pull()

# next, we want to join isoforms that belong to the same gene.
# we import the varroa transcripts ("target_id") and their corresponding gene ("gene_id"). and add a col names.
varroa_isoforms <- read_tsv("/Users/nuriteliash/Documents/GitHub/varroa-virus-networks/data/gene2isoform.txt.gz", col_names = c("gene_id", "target_id"))

# join the varroa transcripts (varroa_isoforms), and the library tpm (df), by the varroa reads ("target_id"). As we do "left_join" to create the final data frame of"gene_tpm" contain ONLY varroa genes, excluding viruses' tpm. 
gene_tpm_71 <-  left_join(varroa_isoforms, df_71, by = "target_id")

#collapse isoforms ("target_id") to a single row of a gene ("gene_id"), and sum the tpm(s) per gene per library
gene_tpm_collps_71 <- gene_tpm_71 %>% 
  gather("library","tpm", -target_id, -gene_id) %>%
  group_by(gene_id, library) %>% 
  summarise(gene_tpm_71 = sum(tpm)) 

# spread the table again, by library
final_gene_tpm_71 <- spread(gene_tpm_collps_71, key = "library", value = "gene_tpm_71") %>% column_to_rownames('gene_id')
  
# transpose final_gene_tpm, and transform (log10+0.000001) 
final_gene_tpm_71_T<- transposeBigData(log10(final_gene_tpm_71 + 0.000001))

PCA_71 <- prcomp(final_gene_tpm_71_T)
p71 <- autoplot(PCA_71, label = TRUE, x = 1, y = 2)+
  ggtitle("PCA of 71 libraries based on gene expression")

# 5 libraries are obvious outliers:"SRR5109825", "SRR5109827", "SRR533974" , "SRR3927496", "SRR8867385".
# when we remove them and repeat the PCA, the library scatter more homogenously:
final_gene_tpm_66_T <- final_gene_tpm_71_T %>%
  rownames_to_column("library") %>%
  dplyr::filter(!(library %in% c("SRR5109825", "SRR5109827", "SRR533974" , "SRR3927496", "SRR8867385"))) %>% column_to_rownames("library")
PCA_66 <- prcomp(final_gene_tpm_66_T)
p66 <- autoplot(PCA_66, label = TRUE, x = 1, y = 2)+
  ggtitle("PCA of 66 libraries based on gene expression")

# plot the two PCA plots together:
grid.arrange(p71,p66,nrow = 1, ncol = 2)

# remove the 5 outlier libraries from the data frame, and save for all subsequent analyses 
df <- df_71 %>% dplyr::select(-c("SRR5109825", "SRR5109827", "SRR533974" , "SRR3927496", "SRR8867385"))
```
saveRDS(df, file = "/data/kallisto.rds")

משום מה - הקוד לא מצליח לראו את כל הוירוסים, רק את הראשון 
dwva


After filtering outlier libraries (based on PCA analysis), we used a total of 66 SRAs belonging to 9 studies of varroa RNAseq for the network analysis (details of the final libraries available in table S5)



## AIM: describing viruses' abundance in the different varroa SRA libraries
first make the suitable data frame containing virus load in each library
```{r first make the suitable data frame containing virus load in each library, message=FALSE, warning=FALSE}
df_old<- read_rds("/Users/nuriteliash/Documents/GitHub/varroa-virus-networks/data/kallisto.rds")
# read the viruses IDs:
virusId <- read_tsv("data/viruses.txt", col_names = c("target_id", "description"))
# rename all 6 VOV_1 segemnts with the same name: "VOV_1"
virusId[3:8, 2] <- "VOV_1"

x <- left_join(virusId, df, by = "target_id") %>% dplyr::select(-target_id) 
  
# VOV_1. virus genome has been described in 6 different segments (Levin et al. 2016). 
# so we sum the 6 segments tpms in each library in a separate table 'VOV_1':
VOV_1 <- x %>% 
      dplyr::slice(3:8) %>%
      group_by(description) %>%
      summarize_all(sum)
#filter out the segments of VOV_1,  
x <- dplyr::filter(x, description != "VOV_1")

# insert VOV_1 in row 3 
r <- 3
insertRow <- function(x, VOV_1, r) {
  x[seq(r+1,nrow(x)+1),] <- x[seq(r,nrow(x)),]
  x[r,] <- VOV_1
  x
}

viruses_load <- insertRow(x, VOV_1, r)
head(viruses_load)

# save the viral load 
saveRDS(viruses_load, file = "/Users/nuriteliash/Documents/GitHub/varroa-virus-networks/results/viruses_load.RData")
```
 
### Vizualize viruses distribution and abundance in (1)abundance plot; (2)PCA of viruses; (3)heat-map
```{r message=FALSE, warning=FALSE}
viruses_load <- readRDS(file = "/Users/nuriteliash/Documents/GitHub/varroa-virus-networks/results/viruses_load.RData")

# (1) virus abundance plot
joined_virus_lib <- viruses_load %>% gather("library", "tpm", -description)

pVirusAbund <- ggplot(joined_virus_lib, aes(x= library, y = tpm, fill = description)) + geom_bar(stat = "identity") + theme(axis.text.x = element_text(angle = 90)) 

# (2) plot a PCA of viruses abundance 
joined_virus_lib_PCA <- joined_virus_lib %>% group_by(description, library) %>% summarise(tpm = sum(tpm)) %>% spread(description, tpm) %>% dplyr::select(-library) 

virus_PCA <- princomp(joined_virus_lib_PCA)  

pVirusPCA <- biplot(virus_PCA)

grid.arrange(pVirusAbund,pVirusPCA,nrow = 1, ncol = 2)

# (3) heat-map of viruses loads (log tpm) per library
# plot the viruses abundance in each library
# make a vector of the new viruses order

new_order <- c("DWVa", "VDV1/DWVb", "VDV2", "BMV","VOV_1","ARV_2", "DWVc", "AmFV", "VDV3", "ABPV", "VTLV", "SV", "BQCV","VDV4", "IAPV", "KBV", "SBPV", "LSV","CBPV" ,"AFV", "ANV", "VPVL_46","VPVL_36")

# re-order the viruses, based on viral load and viral abundance, and change the names of the viruses to match their common name in the literature:
viruses_load_arranged <- viruses_load  %>%
  mutate(description =  factor(description, levels = new_order)) %>%
  arrange(description) %>%
  mutate(across("description", str_replace, "VOV_1", "VOV-1")) %>%
  mutate(across("description", str_replace, "ARV_2", "ARV-2")) %>%
  mutate(across("description", str_replace, "SV", "SBV")) %>%
  mutate(across("description", str_replace, "LSBV", "LSV")) %>%
  mutate(across("description", str_replace, "VDV1/DWVb", "DWVb"))

# lock in factor level order
viruses_load_arranged$description <- factor(viruses_load_arranged$description, levels = viruses_load_arranged$description)

# spread the df with tpm in logscale
virus_abund <- viruses_load_arranged %>%  
    gather("library", "tpm", -description) 

virus_abund[, 3] <- (virus_abund[,3]+ 0.000001)
virus_abund[, 3] <- log10(virus_abund[,3]) 
virus_abund <- replace_with_na(virus_abund, replace = list(tpm = -6))

# now make a heat map
ggplot(data = virus_abund, mapping = aes(x = description,
                                                       y = library,
                                                       fill = tpm)) +
  scale_fill_gradient2(low="#FFFFCC", mid = "#FF9933", high="#990000",  na.value = "grey92") +
  geom_tile() +
  theme_linedraw() +
  xlab(label = "Viruses") +
  ylab(label = "Varroa SRA libraries") +
  labs(fill = "Viral load \n log10(TPM)") + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 15)) + 
  theme(axis.text.y = element_text(size = 12)) + 
  theme(axis.ticks = element_blank()) +
  theme(axis.title = element_text(size = 20))
```

### Virus correlation matrix: 66 libraries, 15 viruses
```{r prepare virus data: 66, message=FALSE, warning=FALSE}
viruses_load <- readRDS(file = "/Users/nuriteliash/OneDrive - OIST/Repos/varroa-virus-networks-Local/results/viruses_load.RData")

# plot the viruses abundance in each library 

# prepare the 'viruses_load' table, with 15 viruses, 
# excluding viruses with 'zero tpm': "CBPV", "AFV", "ANV" , "VPVL_46", "VPVL_36", "SBPV", also excluding viruses with low abundance :"KBV", "LSV".  left with 15 viruses
a_viruses_load_66 <- viruses_load %>%
  filter(description %in% c("DWVa", "ARV_2","VOV_1", "BMV","AmFV","ABPV","VTLV","IAPV","SV","VDV1/DWVb","DWVc","BQCV","VDV3","VDV2","VDV4")) %>%
  column_to_rownames("description") %>% 
  transposeBigData() %>%
  rownames_to_column("library") %>%
  #identify and remove the row numbers of the outlierd libraries (found by PCA) 
  #filter(!(library %in% c("SRR5109825", "SRR5109827", "SRR533974" , "SRR3927496", "SRR8867385"))) %>% 
  column_to_rownames("library")

# correlate viruses load (66 libs)
corrplot2 <- function(data,
                      method = "pearson",
                      sig.level = 0.05,
                      order = "original",
                      diag = FALSE,
                      type = "upper",
                      tl.srt = 75,
                      number.font = 1,
                      number.cex = 0.5,
                      mar = c(0, 0, 0, 0)) {
  
  data_incomplete <- data
  data <- data[complete.cases(data), ]
  mat <- cor(data, method = method)
  cor.mtest <- function(mat, method) {
    mat <- as.matrix(mat)
    n <- ncol(mat)
    p.mat <- matrix(NA, n, n)
    diag(p.mat) <- 0
    for (i in 1:(n - 1)) {
      for (j in (i + 1):n) {
        tmp <- cor.test(mat[, i], mat[, j], method = method)
        p.mat[i, j] <- p.mat[j, i] <- tmp$p.value
      }
    }
    colnames(p.mat) <- rownames(p.mat) <- colnames(mat)
    p.mat
  }
  p.mat <- cor.mtest(data, method = method)
  col <- colorRampPalette(c("#053061","#2166ac","#4393c3","#92c5de","#d1e5f0","#f7f7f7","#fddbc7","#f4a582","#d6604d","#b2182b","#67001f"))
  corrplot(mat,
    method = "color", col = col(200), 
    mar = mar, 
    type = type, order = order,
    tl.col = "black",
    # hide correlation coefficients on the diagonal
    diag = diag
  )
}

virusAbundCor_66 <- corrplot2(
  data = a_viruses_load_66,
  method = "pearson",
  sig.level = 0.05,
  order = "original",
  diag = F, 
  type = "upper",
  tl.srt = 75
)


# correlogram without viruses names:
corrplot2 <- function(data,
                      method = "pearson",
                      order = "original",
                      diag = FALSE,
                      type = "upper",
                      tl.srt = 270,
                      mar = c(0, 0, 0, 0)) {
  
  data_incomplete <- data
  data <- data[complete.cases(data), ]
  mat <- cor(data, method = method)
  cor.mtest <- function(mat, method) {
    mat <- as.matrix(mat)
    n <- ncol(mat)
    p.mat <- matrix(NA, n, n)
    diag(p.mat) <- 0
    for (i in 1:(n - 1)) {
      for (j in (i + 1):n) {
        tmp <- cor.test(mat[, i], mat[, j], method = method)
        p.mat[i, j] <- p.mat[j, i] <- tmp$p.value
      }
    }
    colnames(p.mat) <- rownames(p.mat) <- colnames(mat)
    p.mat
  }
  p.mat <- cor.mtest(data, method = method)
  col <- colorRampPalette(c("#053061","#2166ac","#4393c3","#92c5de","#d1e5f0","#f7f7f7","#fddbc7","#f4a582","#d6604d","#b2182b","#67001f"))
  corrplot(mat,
    method = "color", col = col(200), 
    mar = mar, 
    type = type, order = order,
    tl.col = "black",
    # hide correlation coefficients on the diagonal
    diag = diag
  )
}

virusAbundCor_66 <- corrplot2(
  data = a_viruses_load_66,
  method = "pearson",
 order = "original",
  diag = T, 
 tl.srt = 270,
  type = "upper")

# save the matrix for future analysis (Mantel)
save(virusAbundCor_66, file = "results/virusAbundCor_66.RData")
```

## Gene network analysis
A gene co-expression network identify genes that interact with one another based on common expression profiles (Barabási and Oltvai 2004). Groups of co-expressed genes that have similar expression patterns across samples are identified using hierarchical clustering and are placed in gene ‘modules’ (Miller, Horvath, and Geschwind 2010). 
Weighted gene co-expression analysis was conducted using the WGCNA package in R (Langfelder and Horvath 2008).

